kv/README.md

```diff
@@ -0,0 +1,84 @@
+# Proposed design for JetStream based KV services
+
+This is a proposal to design KV services ontop of JetStream, this requires a KV client to be written, the basic
+features are:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get `string(k)=string(v)` values
+ * Deleting a key completely
+ * Per key TTL
+ * Compacting a key down to n remaining values
+ * Watching a specific key or the entire bucket for live updates
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * read-after-write safety unless the read cache of read replicas were used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
+ * Valid buckets are `^[a-zA-Z0-9_-]+$`
+ * Custom Stream Names and Stream ingest subjects to cater for different domains, mirrors and imports 
+ * Key starting with `_kv` is reserved for internal use
+
+This is an implementation of this concept focussed on the CLI, nats.go and others will have to build language
+specific interfaces focussed on performance and end user.
+
+## Design
+
+### Storage
+
+Given a bucket `CONFIGURATION` we will have:
+
+ * A stream called `KV_CONFIGURATION` with subjects `kv.CONFIGURATION.*`
+ * The stream has Max Messages Per Subject limit set to history with optional placement, R and max age for TTL
+ * Getting a value is an ephemeral consumer with filter subject for the key (`kv.CONFIGURATION.foo`), last received in `Direct` mode
+ * A state of `NumPending+Delivered.Consumer` being 0 means the key does not exist
+ * We store headers as per the table below
+
+### Headers
+
+|Header|Description|
+|------|-----------|
+|KV-Origin-Server|The server name where the client was connected to that created the value|
+|KV-Origin-Cluster|The cluster where the client was connected to that created the value|
+|KV-Origin|If the client opts into this, the IP that created the value|
+
+### Watchers
+
+Watchers can either be per key or per the entire bucket.
+
+For watching a key we simply send key updates over the watch, starting with the latest value or nothing. We will only 
+send the last result for a subject - `NumPending==0`.
+
+For watching the bucket we will send a nil if the bucket is empty, else every result even historical ones.
+
+### Read Replicas
+
+A read replicas a mirror stream from the primary stream.  The KV client is configured to do its reads against the 
```

Comment:
I wonder if it's possible to do something similar with source. 
Each source is it's own key space, that is where the client connects to and performs modifications. 
As a result the writes are spread out.
Then there is an aggregate table using source. it serves as global read cache providing a global view.
Think local device/inventory management at the edge which is aggregated inside HQ for tasks that require a more holistic view. 
Depending on use case replicating deletes might be necessary.

Reply:
come to think of it, how does mirror deal with that?

Reply:
You always write in the main bucket, read locally.  We can't use sources as I care for sequences (could parse sequences off the source headers though but ugh).

It's a interesting optimisation, will keep in mind but I think as is probably this will be fine.  Your most typical operation on KV is simply a thing starting and reading a bunch of values.  This is a thing that can happen entirely regionally.

I think probably for inventory management this just wont work, I do a lot of inventory management work and that's an entirely different use case due to querying needs

Reply:
I will add a `RoKV` interface with only read related functions, when initializing that we can give it a regional as the stream name and that will then work fine, no central access required.  This will be what a lot of uses do primarily I think